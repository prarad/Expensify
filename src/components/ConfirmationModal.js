import React from 'react';
import Modal from 'react-modal';
import { connect } from 'react-redux';
import { closeModal } from '../actions/confirmationModal';
import { startRemoveExpense } from '../actions/expenses';

export const ConfirmationModal = ({
  modalIsOpen,
  handleRequestClose,
  startRemoveExpense,
  selectedExpense = {},
  history
}) => (
    <Modal
      className="modal"
      isOpen={modalIsOpen}
      contentLabel="Confirmation Modal"
      onRequestClose={handleRequestClose}
      closeTime={200}
      ariaHideApp={false}
    >
      <h2 className="modal__title">Confirmation</h2>
      <p className="modal__body">
        Are you sure to delete '{<span>{selectedExpense.description}</span>}' item?
    </p>
      <div className="modal__answers">
        <button
          className="btn"
          onClick={() => {
            startRemoveExpense(selectedExpense.id)
              .then(() => {
                handleRequestClose();
                history.push('/');
              })
          }}
        >Yes</button>
        <button className="btn" onClick={handleRequestClose}>No</button>
      </div>
    </Modal>
  );

const mapStateToProps = state => ({
  modalIsOpen: state.ModalIsOpen
});

const mapDispatchToProps = dispatch => ({
  handleRequestClose: () => dispatch(closeModal()),
  startRemoveExpense: id => dispatch(startRemoveExpense(id))
})

export default connect(mapStateToProps, mapDispatchToProps)(ConfirmationModal);