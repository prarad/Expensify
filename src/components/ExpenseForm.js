import React from 'react';
import { addExpense } from '../actions/expenses'
import moment from 'moment';
import { SingleDatePicker } from 'react-dates';
import 'react-dates/initialize';
import 'react-dates/lib/css/_datepicker.css'

export default class ExpenseForm extends React.Component {
  state = {
    description: '',
    amount: '',
    note: '',
    createdAt: moment(),
    calendarFocused: false,
    error: ''
  }
  dispatch = this.props.dispatch;

  onDescriptionChange = e => {
    const description = e.target.value;
    this.setState(() => ({ description }));
  }
  onAmountChange = e => {
    const amount = e.target.value;
    if (!amount || amount.match(/^\d+(\.\d{0,2})?$/)) {
      this.setState(() => ({ amount }))
    }
  }
  onNoteChange = e => {
    const note = e.target.value;
    this.setState(() => ({ note }))
  }
  onDateChange = createdAt => {
    createdAt &&
      this.setState(() => ({ createdAt }))
  }
  onFocusChange = ({ focused }) => {
    this.setState(() => ({ calendarFocused: focused }))
  }
  onSubmit = e => {
    e.preventDefault();
    if (!this.state.description || !this.state.amount) {
      this.setState(() => ({
        error: 'Please provide description and amount'
      }))
    } else {
      this.setState(() => ({ error: '' }))
    }
    !this.state.error && this.dispatch(
      addExpense({
        description: this.state.description,
        note: this.state.note,
        amount: this.state.amount,
        createdAt: this.state.createdAt
      })
    )
  }
  render() {
    return (
      <div className='Expense-Form'>

        {!!this.state.error && <p>{this.state.error}</p>}
        <form onSubmit={this.onSubmit}>
          <input type="text"
            placeholder='Description'
            onChange={this.onDescriptionChange}
            autoFocus
          />
          <input type="text"
            value={this.state.amount}
            onChange={this.onAmountChange}
            placeholder='Amount'
          />
          <SingleDatePicker
            date={this.state.createdAt}
            onDateChange={this.onDateChange}
            focused={this.state.calendarFocused}
            onFocusChange={this.onFocusChange}
            numberOfMonths={1}
            isOutsideRange={() => false}
          />
          <textarea name="" id="" cols="30" rows="10"
            placeholder='Add a note for your expense(optional)'
            onChange={this.onNoteChange}
          ></textarea>
          <button>Add Expense</button>
        </form>
      </div >
    )
  }
}