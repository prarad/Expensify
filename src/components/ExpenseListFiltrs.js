import React from 'react';
import uuid from 'uuid'
import { connect } from 'react-redux';
import { DateRangePicker } from 'react-dates';
import { setTextFilter, sortByDate, sortByAmount, setEndDate, setStartDate } from '../actions/filters';

export class ExpenseListFilters extends React.Component {
  state = {
    calenderFocused: null
  }
  onDatesChange = ({ startDate, endDate }) => {
    this.props.setStartDate(startDate);
    this.props.setEndDate(endDate);
  }
  onFocusChange = calenderFocused => {
    this.setState(() => ({ calenderFocused }))
  }
  onTextChange = e => {
    this.props.setTextFilter(e.target.value)
  }
  onSortChange = e => {
    e.target.value === 'date' ?
      this.props.sortByDate() :
      this.props.sortByAmount()
  }
  render() {
    return (
      <div className="container">
        <div className="input-group">
          <div className="input-group__item">
            <input
              type="text"
              className="text-input"
              value={this.props.filters.text}
              onChange={this.onTextChange}
              placeholder="Search expenses"
            />
          </div>
          <div className="input-group__item">
            <select
              name="sort-by"
              className="select"
              value={this.props.filters.text}
              value={this.props.filters.sortBy}
              onChange={this.onSortChange}
            >
              <option value="date">Date</option>
              <option value="amount">Amount</option>
            </select>
          </div>
          <div className="input-group__item">
            {this.props.filters.sortBy === 'date' &&
              (
                <DateRangePicker
                  id='date-range-picker'
                  startDate={this.props.filters.startDate}
                  startDateId='start-date-id'
                  endDate={this.props.filters.endDate}
                  endDateId='end-date-id'
                  onDatesChange={this.onDatesChange}
                  focusedInput={this.state.calenderFocused}
                  onFocusChange={this.onFocusChange}
                  showClearDates={true}
                  numberOfMonths={1}
                  isOutsideRange={() => false}
                />
              )
            }
          </div>
        </div>
      </div>
    )
  }
}
const mapStateToProps = state => ({ filters: state.filters })
const mapDispatchToProps = dispatch => ({
  setStartDate: startDate => dispatch(setStartDate(startDate)),
  setEndDate: endDate => dispatch(setEndDate(endDate)),
  setTextFilter: text => dispatch(setTextFilter(text)),
  sortByDate: () => dispatch(sortByDate()),
  sortByAmount: () => dispatch(sortByAmount())
})

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters)