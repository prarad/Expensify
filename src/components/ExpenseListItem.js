import React from 'react';
import { Link } from 'react-router-dom';
import moment from 'moment';
import numeral from 'numeral';

export default ({ id, description, amount, createdAt }) => (
  <div className='Item'>
    <Link to={`edit/${id}`} >
      <h2>{description}</h2>
    </Link>
    <p>{numeral(amount / 100).format('$0,0.00')}</p>
    {createdAt &&
      <p>{
        moment(createdAt).format('MMMM Do, YYYY')
      }</p>
    }
  </div>
)