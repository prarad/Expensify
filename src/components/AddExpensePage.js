import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { addExpense } from '../actions/expenses';

const AddExpensePage = ({ dispatch, history }) => (
  <div className='AddExpensePage'>
    <h1>Add Expense</h1>
    <ExpenseForm
      action='add'
      onSubmit={(expense => {
        dispatch(addExpense(expense))
        history.push('/')
      })} />
  </div>
)

export default connect()(AddExpensePage)