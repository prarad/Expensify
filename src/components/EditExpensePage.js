import React from 'react';
import { connect } from 'react-redux';
import { startEditExpense, startRemoveExpense } from '../actions/expenses';
import ExpenseForm from './ExpenseForm';
import ConfirmationModal from '../components/ConfirmationModal';
import { openConfirmationModal } from '../actions/confirmationModal';

export class EditExpensePage extends React.Component {
  onSubmit = (expense) => {
    this.props.startEditExpense({
      id: this.props.expense.id,
      updates: expense
    });
    this.props.history.push('/')
  }
  onRemove = () => {
    this.props.openModal()
  }
  render() {
    return (
      < div >
        <div className="page-header">
          <div className="container">
            <h1 className="page-header__title">Edit Expense</h1>
          </div>
        </div>
        <div className="container">
          <ExpenseForm
            expense={this.props.expense}
            onSubmit={this.onSubmit}
          />
          <ConfirmationModal selectedExpense={this.props.expense} history={this.props.history} />
          <button
            className="btn btn--secondary"
            onClick={this.onRemove}
          >Remove Expense</button>
        </div>
      </div >
    )
  }
}

const mapDispatchToProps = (dispatch, props) => ({
  openModal: () => dispatch(opennModal()),
  startEditExpense: ({ id, updates }) => dispatch(startEditExpense({ id, updates }))
})

const mapStateToProps = (state, props) => ({
  expense: state.expenses.find(expense => expense.id === props.match.params.id)
})

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage)