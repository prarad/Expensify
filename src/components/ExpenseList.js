import React from 'react';
import { connect } from 'react-redux';
import ExpenseListItem from './ExpenseListItem';
import getVisibleExpenses from '../selectors/expenses'

export const ExpenseList = props => (
    <div>
        <h1>Expense List</h1>
        {props.expenses.map(expense => (
            <ExpenseListItem {...expense} key={expense.id} />
        ))}
    </div>
)
const mapStateToProps = state => ({
    expenses: getVisibleExpenses(state.expenses, state.filters)
})

export default connect(mapStateToProps)(ExpenseList)