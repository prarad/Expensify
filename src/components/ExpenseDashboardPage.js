import React from 'react';
import ExpenseList from './ExpenseList';
import ExpenseListFilters from './ExpenseListFiltrs';

export default () => (
  <div className="ExpenseDashboardPage">
    <ExpenseListFilters />
    <ExpenseList />
  </div>
)