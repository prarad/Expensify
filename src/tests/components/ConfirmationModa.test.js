import React from 'react';
import { shallow } from 'enzyme';
import { ConfirmationModal } from '../../components/ConfirmationModal';

test('should correctly render confirmation modal', () => {
  const wrapper = shallow(<ConfirmationModal />);
  expect(wrapper).toMatchSnapshot()
})