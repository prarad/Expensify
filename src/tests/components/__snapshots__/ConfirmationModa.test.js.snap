// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should correctly render confirmation modal 1`] = `
<Modal
  ariaHideApp={false}
  bodyOpenClassName="ReactModal__Body--open"
  className="modal"
  closeTime={200}
  closeTimeoutMS={0}
  contentLabel="Confirmation Modal"
  isOpen={false}
  parentSelector={[Function]}
  portalClassName="ReactModalPortal"
  shouldCloseOnEsc={true}
  shouldCloseOnOverlayClick={true}
  shouldFocusAfterRender={true}
  shouldReturnFocusAfterClose={true}
>
  <h2
    className="modal__title"
  >
    Confirmation
  </h2>
  <p
    className="modal__body"
  >
    Are you sure to delete '
    <span />
    ' item?
  </p>
  <div
    className="modal__answers"
  >
    <button
      className="btn"
      onClick={[Function]}
    >
      Yes
    </button>
    <button
      className="btn"
    >
      No
    </button>
  </div>
</Modal>
`;
